api/ondemand_vhds_rds_protocol.rst

```diff
@@ -0,0 +1,187 @@
+in the mesh. This causes scaling issues as the size of the cluster grows. The
+majority of this complexity can be found in the virtual host configurations, of
+which most are not needed by any individual proxy. With a goal of being able
+to scale to one million vhosts in the future, the capability to filter a
+proxy to only contain the required virtual hosts is a necessity.
+
+In order to fix this issue, we are implementing on-demand RDS. On-demand RDS,
+uses the delta xDS protocol, which is a separate protocol xDS. Instead of
+always sending the full route config, Envoy will be able to subscribe/
+unsubscribe from a list of resources stored internally in the xDS management
+server. The xDS management server will monitor this list and use it to filter
+the configuration sent to an individual Envoy instance to only contain the
+subscribed resources.
+
+Joining/Reconnecting
+====================
+When an Envoy instance forms a delta xDS connection to the xDS management
+server for the first time, it will receive a base configuration filtered down
+to a subset of routes that are likely useful to the Envoy instance. For
+example, these might be filtered down by routes that exist in the current
+namespace. The xDS management server will send this same collection of
+resources on reconnect.
+
+.. figure:: diagrams/delta_rds_connection.svg
+   :alt: Delta RDS connection
+
+Requesting Additional resources
+===============================
+In order to request additional resources, Envoy will send a
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+to a xDS management server. The xDS management server will then respond with a
+:ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`
+which is followed up by an Ack/Nack
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+from Envoy.
+
+.. figure:: diagrams/delta_rds_request_additional_resources.svg
+   :alt: Delta RDS request additional resources
+
+Virtual Host Discovery Service
+==============================
+In order to support on-demand, an additional protocol, VHDS, will be added.
+This protocol allows a separation of concerns with RDS where RDS is in charge
+of maintaining route configs while VHDS is in charge of communicating virtual
+hosts. In VHDS, Envoy will send a
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+with type_url set to type.googleapis.com/envoy.api.v2.route.VirtualHost and
+resource_names set to a list of routeconfig names + the request HTTP "host"
+headers for which it would like configuration. The management server will
+respond with a
+:ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`
+with the resources field populated with the specified virtual hosts, the name
+field populated with the virtual host name, and the alias field populated with
+the explicit (no wildcard) domains of the virtual host. Future updates to these
```

Comment:
What if a virtual host is just `*.foo.com`? Is the alias set empty?

Reply:
At the very least the alias field must contain the value of the host/authority header that was used in the on-demand request. Otherwise it will be impossible to match DeltaDiscoveryResponses with DeltaDiscoveryRequests.

Reply:
The alias field must always contain the host/authority header that was used in the on-demand request (just the header without the route_config_name prefix), even if the management server couldn't resolve it: this is used to match paused connections/filter chains to received DeltaDiscoveryResponses.